teps/0089-spire-support.md

```diff
@@ -0,0 +1,215 @@
+2. Tekton Chains has no way to verify that the TaskRun it received wasn't modified by anybody other than Tekton, during or after execution
+3. Tekton Pipelines can't verify that the results it reads weren't modified
+
+This is where SPIRE comes in!
+SPIRE can be used to request signatures and certificates (SVIDs) for a given workload (this workload could be a result, or in our case an entire TaskRun yaml).
+SPIRE can be configured to only issue SVIDs to certain k8s workloads with a SPIRE [k8s Workload Attestor](https://github.com/spiffe/spire/blob/main/doc/plugin_agent_workloadattestor_k8s.md).
+In thise case, we'll set up SPIRE to only issue SVIDs to:
+* The Tekton controller
+* Pods created by the Tekton controller (e.g. pods running the entrypointer image)
+
+This configuration will let us ensure that all Tasks were created by Tekton and that Task results weren't modified.
+
+The basic flow would look like this:
+1. Tekton Pipelines receives a TaskRun config, and generates the Pod for with SPIRE mounted in to the Pod
+1. Tekton Pipelines stores the state of the TaskRun in memory, and every time the status updates it follows this loop:
+* Compare the current TaskRun to the state stored in memory
+* If the TaskRun has been changed, don't request an SVID from SPIRE
+* If the TaskRun has not been changed, then update the TaskRun status and store the new state in memory
+1. The entrypointer image requests an SVID and signature for all results in the Pod
+1. Tekton Pipelines verifies the results and stops if verification fails
+1. If successful, it requests a signature over the completed TaskRun yaml from SPIRE
+1. Tekton Pipelines stores the SVID and signature received from SPIRE as annotations on the TaskRun
+1. Tekton Chains verifies that SVID and signature against the TaskRun yaml
+1. If verification is successful, Chains proceeds normally. Otherwise, it stops and doesn't sign anything!
+
+With the SPIRE integration, we successfully mitigate the two issues mentioned above.
+Tekton Chains can be confident the TaskRun was created by Tekton, since only Tekton workloads can have SPIRE issued SVIDs.
+Tekton Chains can also verify that the TaskRun yaml wasn't modified by anybody other than the Tekton controller, because the signature exists over the entire TaskRun yaml.
+
+_NOTE: A cluster administrator could still potentially threaten this model by disabling SPIRE or changing settings on the k8s Workload Attestor. At SLSA 3, it is still allowed that some trusted administrators will have root access to the build system. This will eventually need to be addressed for SLSA 4_.
+
+
+## Motivation
+
+Security! =)
+
+We'll also need this feature for Tekton to achieve [SLSA 3](https://slsa.dev/levels), which requires Non-falsifiable Provenance.
+
+### Requirements
+* SPIRE is configured to only issue SVIDs to the Tekton controller via a [k8s Workload Attestor](https://github.com/spiffe/spire/blob/main/doc/plugin_agent_workloadattestor_k8s.md)
```

Comment:
i think we might want to include some requirements around supporting setups where it is expected that TaskRuns might be modified by something other than the Pipelines Controller, e.g.:

* A more common case where labels and annotations might be updated by other services
* A less common case where mutating admission controllers are intentionally changing the TaskRun (e.g. injecting steps - maybe we don't need to consider this at the level of the TaskRun tho, but if we want to extend this approach to PipelineRuns I think we would - e.g. SolarWinds is injecting tasks into Pipeline execution to do things like image scans)

Reply:
I added these into a list of points to consider for when we thoroughly design Signed Taskruns!